scaladoc/src/dotty/tools/scaladoc/tasty/comments/Comments.scala

```diff
@@ -129,16 +129,12 @@ abstract class MarkupConversion[T](val repr: Repr)(using dctx: DocContext) {
       val path = s.source.map(_.path)
       val pathBasedArg = dctx.snippetCompilerArgs.get(path)
       val data = SnippetCompilerDataCollector[qctx.type](qctx).getSnippetCompilerData(s, s)
+      val sourceFile: dotty.tools.dotc.util.SourceFile = s.pos.fold(dotty.tools.dotc.util.NoSource)(_.sourceFile.asInstanceOf[dotty.tools.dotc.util.SourceFile])
```

Comment:
Can we get rid of this cast as well? How should we handle failures?

Reply:
Yes, I'll refactor this cast. In 99.99% cases there will be source file associated with symbol, especially in end-users projects. Maybe we should just produce warning/info with information that snippet compiler couldn't obtain source files for certain snippets and due to that the positions might not be accurate?

Reply:
In which situation would there be no files?

Reply:
I'm not so fluent with the compiler internals but I can guess that there might be a problem with symbols created synthetically in compiler? For some reason the Position in Symbol is optional, maybe @smarter can provide there some information.

Reply:
I don't know the context here, if it's a symbol coming from sources it should always have a source associated with it, but you can use `s.source` to access that just like you do above instead of going through `s.pos`.

Reply:
`s.source` is available only using compiler API of symbol but we need to cast between Quotes API and Compiler API either way. I think that smarter's solution is better because it's used in some places of Scaladoc and seems stable. This way we get `SourceFile` instead of `Option[SourceFile]` but still we need to handle `NoSource`. Actually, getting `NoSource` is not a failure and should be supported.

Reply:
Oh I didn't realize it wasn't available in the quotes API, yeah we should try to uniformize these things.

>  but we need to cast between Quotes API and Compiler API 

This is something we absolutely need to get rid of in the future (scaladoc should only use the public API, anything missing in the public API needs to be added to it), so I would caution against making the problem worse by doing even more casting.

Reply:
99% of casting is done in file called SyntheticSupport, maybe we should take a look at it and create an issue containing all things missing in public API? That was actually a purpose of this file

Reply:
> create an issue containing all things missing in public API?

sounds good yes